source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -25,48 +25,60 @@ function isActionAllowed(action: Action, forbiddenActionTypes: Array<string>): b
     return true;
 }
 
+enum InstanceCreationErrorType {
+    UnknownType,
+    ForbiddenType
```

Comment:
Neat. Does this get bubbled up to the author or is this enforced host-side? Assuming it gets bubbled up, is there an equivalent mapping of error type to error message (user-friendly string) that can get bubbled up/included as well?

Reply:
Not 100% sure what you're saying. These errors get bubbled up to the developer (when they integrate the AC renderer in their own app) to help them debug and make runtime decisions. These errors are also bubble up in the designer to help the user "debug" (correct) their card. There isn't a single string associated with each error code; the string is contextual, the code is there to help the calling code make decisions.

Reply:
Speaking of error message to go with the error code; For eg ForbiddenType by itself isnt that helpful as an associated error string with it that could be more descriptive for eg. I think though this is a larger ticket cleanup item across all our renderers (consolidating on error codes + messages as well as tackling localization of those error messages) at some point. Doesnt need to block this PR

---
In reply to: [279439868](https://github.com/Microsoft/AdaptiveCards/pull/2771#discussion_r279439868) [](ancestors = 279439868)

Reply:
There is always a string associated with the error code. The string is for debug purposes and is in most case very descriptive of what the problem is. The error code is what allows an app to make a decision at runtime as to whether it should display the card. ForbiddenType means the parser encountered an action or element type that isn't allowed. This might have caused fallback to happen or the action/element being dropped. An app can decide that in those situations it doesn't even want to render the card.